### PR TITLE
[debug] Add auto crash dumps for Windows

### PR DESCRIPTION
**Pull request by zach2good**
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [x] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

Finally starting to split out debug stuff by platform. This was working last night, but it isn't working now. So, it'll be ready to go once I test it and start producing dumps : poop:

### CODE REVIEW

**Comment by zach2good** [comment]
I've set this up to create dumps in the root of the repo now:
- A: Because I can't figure out how to make `CreateFileA` create a file inside a child directory
and
- B: It makes it more obvious to the user that there has been a crash, and that there are dumps available 🤷

**Comment by zach2good** [comment]
Maybe I should hide this behind a setting?

**Comment by TeoTwawki** [comment]
>     * A: Because I can't figure out how to make `CreateFileA` create a file inside a child directory
>       and

what have you tried?

**Comment by zach2good** [comment]
`dmp/<name>` and `dmp\\<name>` both didn't produce anything

**Comment by TeoTwawki** [comment]
> `dmp/<name>` and `dmp\\<name>` both didn't produce anything

iirc I had to prefix the leading self ref and slash for the directory. like `.\\dmp\\filename`  last time I was configuring a path with CreateFileA  (it has been a few yrs I think though). Other current projects just use `MoveFileA` right after instead and I forget why that was preferred.

**Comment by zach2good** [comment]
I'll test it out tomorrow 👍

**Comment by TeoTwawki** [comment]
Even if that doesn't work out at least repo root is better than directing ppl to look for the default windows location - most wouldn't know where that is. The windows registry can configure dump locations per exe when WER handles them but the auto saving with that has been broken for a long time now.
ex:
```reg
[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\Windows Error Reporting\LocalDumps\name.exe]
"DontShowUI"=dword:00000001
"DumpFolder"=hex(2):43,00,3a,00,5c,00,43,00,72,00,61,00,73,00,68,00,44,00,75,\
  00,6d,00,70,00,5c,00,47,00,61,00,6d,00,65,00,53,00,65,00,72,00,76,00,65,00,\
  72,00,5c,00,4d,00,61,00,69,00,6e,00,4d,00,61,00,70,00,00,00
"DumpCount"=dword:00000014
"DumpType"=dword:00000002
"CustomDumpFlags"=dword:00000000
```
iunno why it wants folder path in hex. its less than an ideal method of handling crashes. I'll surrender configurable paths if it gets automated saving. :+1:

**Comment by zach2good** [comment]
I figured out the problem, I was using the results of `GetModuleFileNameA`, which was a full file path. Everything is working as expected now.

**Comment by zach2good** [comment]
Ready for review please 🙏

**Review by claywar — APPROVED** [review]

File: src/common/utils.cpp [hunk 1]
```diff
@@ -729,3 +729,9 @@ bool definitelyLessThan(float a, float b)
     constexpr float epsilon = std::numeric_limits<float>::epsilon();
     return (b - a) > ((fabs(a) < fabs(b) ? fabs(b) : fabs(a)) * epsilon);
 }
+
+void crash()
+{
+    // cppcheck-suppress nullPointer
+    *((unsigned int*)0) = 0xDEAD;
```
Comment (TeoTwawki) [inline src/common/utils.cpp:736]: haha I see what you did there

**Review by TeoTwawki — DISMISSED** [review]

**Comment by zach2good** [comment]
Don't merge yet, I want to experiment with adding frame locals to the stack trace in this PR:
https://accu.org/journals/overload/29/165/orr/

**Comment by zach2good** [comment]
Moving towards this sort of format for crash logging:
```
[01/02/22 17:36:36:905][critical][stacktrace] ================================================================
[01/02/22 17:36:36:905][critical][stacktrace] !!! CRASH !!!
[01/02/22 17:36:36:905][critical][stacktrace] Exception STATUS_ACCESS_VIOLATION (0XC0000005) occured!
[01/02/22 17:36:36:905][critical][stacktrace] A crash dump has been created: <repo root>/dmp/topaz_game_64.exe_2022_01_02_17_36_36.dmp
[01/02/22 17:36:36:905][critical][stacktrace] ================================================================
[01/02/22 17:36:37:081][critical][stacktrace] at crash in C:\ffxi\server\src\common\utils.cpp: line: 736: address: 0x7FF7DF67FA40
[01/02/22 17:36:37:082][critical][stacktrace] at zoneutils::LoadNPCList in C:\ffxi\server\src\map\utils\zoneutils.cpp: line: 307: address: 0x7FF7DFA2C110
[01/02/22 17:36:37:106][critical][stacktrace]   L std::string(32) testString: Hello World!
[01/02/22 17:36:37:106][critical][stacktrace]   L *std::string(8) testRef: ?
[01/02/22 17:36:37:106][critical][stacktrace]   L *CNpcEntity(8) PNpc: ?
[01/02/22 17:36:37:106][critical][stacktrace]   L unsigned int(4) NpcID: 16781331 (0x1001013)
[01/02/22 17:36:37:106][critical][stacktrace]   L *char(8) contentTag: ìü▓^±☺
[01/02/22 17:36:37:106][critical][stacktrace]   L unsigned int(2) ZoneID: 1
[01/02/22 17:36:37:106][critical][stacktrace]   L int(4) ret: 0 (0x0)
[01/02/22 17:36:37:106][critical][stacktrace]   L *char(8) Query: Çþ╝▀¸
[01/02/22 17:36:37:106][critical][stacktrace]   L [22](22) address: 0.0.0.0
[01/02/22 17:36:37:106][critical][stacktrace] at zoneutils::LoadZoneList in C:\ffxi\server\src\map\utils\zoneutils.cpp: line: 686: address: 0x7FF7DFA2F660
[01/02/22 17:36:37:106][critical][stacktrace]   L *char(8) query: pÕ╝▀¸
[01/02/22 17:36:37:106][critical][stacktrace]   L int(4) ret: 0 (0x0)
[01/02/22 17:36:37:106][critical][stacktrace]   L std::vector<unsigned short,std::allocator<unsigned short> >(24) zones: ?
[01/02/22 17:36:37:106][critical][stacktrace]   L [22](22) address: 0.0.0.0
[01/02/22 17:36:37:107][critical][stacktrace] at do_init in C:\ffxi\server\src\map\map.cpp: line: 256: address: 0x7FF7DF5C4A00
[01/02/22 17:36:37:107][critical][stacktrace]   P int(4) argc: 1 (0x1)
[01/02/22 17:36:37:107][critical][stacktrace]   P **char(8) argv: ?
[01/02/22 17:36:37:107][critical][stacktrace]   L *_iobuf(8) SETTINGS_MAIN: ?
[01/02/22 17:36:37:107][critical][stacktrace] at main in C:\ffxi\server\src\common\kernel.cpp: line: 265: address: 0x7FF7DF68A1C0
[01/02/22 17:36:37:107][critical][stacktrace]   L std::chrono::duration<__int64,std::ratio<1,10000000> >(8) next: ?
[01/02/22 17:36:37:107][critical][stacktrace]   P int(4) argc: 1 (0x1)
[01/02/22 17:36:37:107][critical][stacktrace]   P **char(8) argv: ?
[01/02/22 17:36:37:107][critical][stacktrace]   L fd_set(8008) rfd: ?
[01/02/22 17:36:37:107][critical][stacktrace] ================================================================
[01/02/22 17:36:37:107][critical][stacktrace] Server Name: X
[01/02/22 17:36:37:107][critical][stacktrace] Players Online: X
[01/02/22 17:36:37:107][critical][stacktrace] OS Version: X
[01/02/22 17:36:37:107][critical][stacktrace] Build Date/Time: X
[01/02/22 17:36:37:107][critical][stacktrace] Process Name: topaz_game_64.exe
[01/02/22 17:36:37:107][critical][stacktrace] Process Uptime: X
[01/02/22 17:36:37:107][critical][stacktrace] Git SHA: X
[01/02/22 17:36:37:107][critical][stacktrace] Git Message: X
[01/02/22 17:36:37:107][critical][stacktrace] ================================================================

[process exited with code 1]
```

Still need to figure out how to iterate through user type pointers and pull our the names, ids and locations of entities, but it's starting to come together

**Comment by zach2good** [comment]
Ok, I think I'm happy for this to go in. I've not added settings yet for configuring how many memory dumps you can hold, if you're going to auto generate them etc. I want to do that as part of the future PR where I tear apart all of the settings.

**Comment by zach2good** [comment]
The default behaviour is:
Create memory dumps: YES
Max dumps in dump dir: 3

**Review by claywar — APPROVED** [review]

**Pull request merged**